### PR TITLE
Filtering skeleton

### PR DESCRIPTION
**Pull request by CremaLuca**
## Premise
This PR aims to implement partially what has been discussed about this [Trello board](https://trello.com/c/TY4yCJPi/54-filtering-pattern)

## Features

Some untold rules:

- When a stream is flagged as `is_finished` no atoms shall be added to it.
- A filter is finished when all of its outputs are flagged as `is_finished`

#### feat(`stream.py`)
It's a list of atoms that uses a `StreamIterator` as default iterator.
**+** StreamIterator has a `has_next()` method and when calling `next()` it returns an already removed atom from the stream.

#### feat(`filter.py`)
It's an abstract class for filters, defines the inputs and the outputs of the filter, the parameters of the `execute()` method and the `is_finished()` method.

#### feat(`filter_layer.py`)
Defines a set of filters, that's it, nothing more.

#### feat(`filter_list.py`)
It's the sequence of filter layers. The class defines the behaviour of execution of the filters.

#### feat(`utils/time_handler.py`)
Utils class to avoid importing datetime everytime we need to work on datetimes, should be updated on later PR's to include operations done in the `downloader` classes.

#### test(`utils/time_handler.py`)
**+** Test for `str_to_datetime`
**+** Test for `datetime_to_str`

### CODE REVIEW

File: otri/filtering/stream.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    def __init__(self, iterable : Iterable):
+        self.iterable = iterable
+
+    def __next__(self):
+        '''
+        Pops the first element of the given collection.
+        '''
+        value = self.iterable[0]
+        del self.iterable[0]
```
Comment (ricdezen) [inline otri/filtering/stream.py]: Funge? Toglierlo lo toglie, ma lo cancella anche dalla memoria penso (???)

Comment (CremaLuca) [reply otri/filtering/stream.py]: È quello che vogliamo, vogliamo liberare la memoria dello stream attuale una volta che è stato letto un atomo, se ne occuperà il filtro a ficcare l'atomo in un altro stream per tenerlo vivo (o distruggerlo perché non serve)

Comment (ricdezen) [reply otri/filtering/stream.py]: Il punto è che così distruggi secco l'atomo, il che assume che il ricevente ne abbia fatto una deep copy, è necessario? La garbage collection lo rimuoverebbe ugualmente se non fosse più usato

Comment (CremaLuca) [reply otri/filtering/stream.py]: Il garbage collector non lo rimuoverebbe finché la lista non viene più usata, se dobbiamo processare 1TB di atomi e abbiamo tre filtri alla fine ci troveremo con 3TB di stream che contengono i tre stadi di ogni atomo, invece vogliamo che quel 1TB rimanga uno solo

Comment (ricdezen) [reply otri/filtering/stream.py]: Ok provando ho scoperto che un del su un elemento della lista lo rimuove solo, ero convinto che avrebbe cancellato anche l'oggetto. Non lo fa perchè gli elementi di una lista sono oggetti a parte, dei puntatori.

File: otri/filtering/filters/interpolation_filter.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        if(interval == "days"):
+            value = diff.days() - 1
+        if(value == None):
+            raise ValueError("Interval {} is not supported".format(interval))
+        if(value < 0):
+            raise ValueError(
+                "The atom list is not datetime-ordered or two atoms have the same datetime:\n{}\n{}".format(atom1, atom2))
+        return value
+
+    def __timedelta_from_interval(self, interval: str):
```
Comment (ricdezen) [inline otri/filtering/filters/interpolation_filter.py]: A dict sounds easier

File: otri/filtering/filters/summer_filter.py [hunk 3]
```diff
@@ -10,21 +10,21 @@ class SummerFilter(Filter):
         Single stream
     '''
 
-    def __init__(self, input_stream: Stream, keys_to_change: Collection[str], const : int):
+    def __init__(self, input_stream: Stream, keys_to_change: Collection[str], constants : Mapping):
```
Comment (ricdezen) [inline otri/filtering/filters/summer_filter.py]: Can directly be a mapping of the keys to their constants

Comment (CremaLuca) [reply otri/filtering/filters/summer_filter.py]: Sure

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- otri/filtering/filters/math_filter.py  5
- otri/filtering/filters/generic_filter.py  4
- otri/filtering/filters/merge_filter.py  6
- test.py  1
- autocorrelation.py  3
- test/filtering/filter_test.py  2
- test/filtering/filters/nuplicator_filter_test.py  2
- otri/filtering/filters/phase_filter.py  6
- otri/utils/time_handler.py  1
- test/filtering/filters/generic_filter_test.py  3
- test/utils/time_handler_test.py  1
- otri/filtering/filters/interpolation_filter.py  6
- otri/filtering/filter.py  3
- otri/filtering/filter_list.py  13
- test/filtering/filters/merge_filter_test.py  1
- otri/database/database_stream.py  1
- otri/filtering/filters/nuplicator_filter.py  6
- otri/filtering/filter_layer.py  3
- otri/filtering/stream.py  2
- test/filtering/filters/phase_filter_test.py  2
- otri/filtering/filters/statistics_filter.py  6
- test/filtering/stream_test.py  1
- test/filtering/filters/statistics_filter_test.py  2
         

Clones added
============
- autocorrelation.py  2
- test/filtering/filters/merge_filter_test.py  2
- test/filtering/filters/phase_filter_test.py  5
- test/filtering/filters/statistics_filter_test.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939&bid=17990617)

File: otri/filtering/filters/generic_filter.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+from ..filter import Filter, Stream, Collection
```
Comment (CremaLuca) [inline otri/filtering/filters/generic_filter.py:1]: Codacy Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/statistics_filter.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
+from ..filter import Filter, Stream, Collection
+from typing import Sequence, Mapping, Dict
```
Comment (CremaLuca) [inline otri/filtering/filters/statistics_filter.py:2]: Codacy Issue found: [Unused Dict imported from typing](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: test.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
+from otri.filtering.filter_list import FilterList, FilterLayer
+from otri.filtering.stream import Stream
+from otri.filtering.filters.interpolation_filter import InterpolationFilter
+from otri.filtering.filters.multiplier_filter import MultiplierFilter
+from otri.filtering.filters.math_filter import MathFilter
+from otri.filtering.filters.average_filter import AverageFilter
+from otri.downloader.yahoo_downloader import YahooDownloader, date
+from otri.downloader.alphavantage_downloader import AVDownloader
```
Comment (CremaLuca) [inline test.py:8]: Codacy Issue found: [Unused AVDownloader imported from otri.downloader.alphavantage_downloader](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/nuplicator_filter.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+from ..filter import Filter, Stream, Collection
```
Comment (CremaLuca) [inline otri/filtering/filters/nuplicator_filter.py:1]: Codacy Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/statistics_filter.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+from numbers import Number
```
Comment (CremaLuca) [inline otri/filtering/filters/statistics_filter.py:3]: Codacy Issue found: [Unused Number imported from numbers](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: autocorrelation.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    f_list_1.execute()
+
+    # Filter list 2
+
+    normalize_filter = MathFilter(
+        input_stream=stats_filter.get_output_stream(0),
+        keys_operations={k: lambda value: value/v
+                         for k, v in stats_filter.get_max().items()}
+    )
+    f_layer_normalize = FilterLayer([normalize_filter])
```
Comment (CremaLuca) [inline autocorrelation.py:79]: Codacy Issue found: [Unused variable 'f_layer_normalize'](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/math_filter.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
+from ..filter import Filter, Stream, Collection
```
Comment (CremaLuca) [inline otri/filtering/filters/math_filter.py:1]: Codacy Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- otri/filtering/filter_layer.py  3
- otri/filtering/filters/math_filter.py  5
- otri/utils/time_handler.py  1
- test/filtering/filter_test.py  2
- test.py  1
- test/filtering/filters/generic_filter_test.py  3
- test/filtering/filters/nuplicator_filter_test.py  2
- otri/filtering/filter.py  3
- autocorrelation.py  3
- otri/filtering/filters/nuplicator_filter.py  6
- test/filtering/filters/statistics_filter_test.py  2
- test/filtering/filters/phase_filter_test.py  2
- otri/filtering/filters/statistics_filter.py  6
- otri/filtering/filters/phase_filter.py  6
- test/filtering/stream_test.py  1
- test/utils/time_handler_test.py  1
- otri/filtering/filters/merge_filter.py  6
- otri/filtering/filter_list.py  13
- otri/filtering/filters/interpolation_filter.py  6
- otri/filtering/filters/generic_filter.py  4
- otri/database/database_stream.py  1
- test/filtering/filters/merge_filter_test.py  1
- otri/filtering/stream.py  2
         

Clones added
============
- filter_styles.py  2
- autocorrelation.py  2
- test/filtering/filters/statistics_filter_test.py  7
- test/filtering/filters/phase_filter_test.py  5
- test/filtering/filters/merge_filter_test.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939&bid=17990617)

File: autocorrelation.py (hunk 9 above)
Comment (CremaLuca) [inline autocorrelation.py:79]: Codacy Issue found: [Unused variable 'f_layer_normalize'](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/generic_filter.py (hunk 4 above)
Comment (CremaLuca) [inline otri/filtering/filters/generic_filter.py:1]: Codacy Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/math_filter.py (hunk 10 above)
Comment (CremaLuca) [inline otri/filtering/filters/math_filter.py:1]: Codacy Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: test.py (hunk 6 above)
Comment (CremaLuca) [inline test.py:8]: Codacy Issue found: [Unused AVDownloader imported from otri.downloader.alphavantage_downloader](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/statistics_filter.py (hunk 8 above)
Comment (CremaLuca) [inline otri/filtering/filters/statistics_filter.py:3]: Codacy Issue found: [Unused Number imported from numbers](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: filter_styles.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+            ],
+            policy=BACK_IF_NO_OUTPUT
+        ),
+        FilterLayer(
+            filters=[
+                # Delta calculation
+                PhaseDeltaFilter(
+                    input="ticker1_interp",
+                    output="ticker1_delta"
+                    keys=KEYS_TO_CHANGE,
```
Comment (CremaLuca) [inline filter_styles.py:57]: Codacy Issue found: [Syntax Error (MC0000)](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/statistics_filter.py (hunk 5 above)
Comment (CremaLuca) [inline otri/filtering/filters/statistics_filter.py:2]: Codacy Issue found: [Unused Dict imported from typing](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: otri/filtering/filters/nuplicator_filter.py (hunk 7 above)
Comment (CremaLuca) [inline otri/filtering/filters/nuplicator_filter.py:1]: Codacy Issue found: [Unused Collection imported from filter](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5489939)

File: autocorrelation.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
+"""
+Module to calculate autocorrelation within a single ticker.
+"""
+
+__autor__ = "Riccardo De Zen <ricdezern@gmail.com>, Luca Crema <lc.crema@hotmail.com>"
```
Comment (ricdezen) [inline autocorrelation.py]:
```suggestion
__autor__ = "Riccardo De Zen <riccardodezen98@gmail.com>, Luca Crema <lc.crema@hotmail.com>"
```

**Pull request merged**